### PR TITLE
PANTHER GO

### PR DESCRIPTION
**Pull request by typhainepl**
Example of signatures used: 
signatures/PTHR23315/PTHR12120/PR01969/go/#!

### CODE REVIEW

**Review by matthiasblum — COMMENTED** [review]
What is the aim of this PR? Because I am not sure I understand it.

The new value presented for a given PANTHER family is the number of its subfamilies associated with a given GO term, and not a number of proteins. I find this slightly confusing.

**Comment by typhainepl** [comment]
@matthiasblum up to date databases for testing are IPDEV and intdev.
Shouldn't be deployed before pyinterprod pipeline is ran again.

File: pronto/api/signatures/go.py [hunk 1]
```diff
@@ -95,3 +99,62 @@ def get_go_terms(accessions):
...
+
+        try:
+            term = terms[term_id]
+        except KeyError:
+            term = terms[term_id] = get_go_details(term_id, pg_cur, aspects_stmt, params)
+        
+        try:
+            s = term["signatures"][pth_fam]
+        except KeyError:
+            s = term["signatures"][pth_fam] = [set(), set(), set()]
```
[the remark this answers is not in the document]
Comment (typhainepl) [reply pronto/api/signatures/go.py]: I've added a check, but open to suggestions

**Review by matthiasblum — CHANGES_REQUESTED** [review]
Please find comments below. 

The only issue I have is the use of string interpolations in SQL statements instead of using bind variables, because this is a security issue.

There are a few things where performance could become an issue, but that would be for PANTHER families matching a lot of proteins.

File: pronto/static/js/signatures/go.js [hunk 2]
```diff
@@ -21,22 +21,38 @@ function getGoTerms(accessions) {
...
+                        <th rowspan="2">${data.results.length.toLocaleString()} terms</th>
+                        <th class="collapsing center aligned" rowspan="2"><button class="ui primary small fluid compact icon button"><i class="sitemap icon"></i></button></th>
                         ${accessions.map(genCell).join('')}
                         </tr>
-                        </thead>`;
+                        <tr>`;
+
+            for (let i = 0; i < accessions.length; i++) {
+                html += `
+                        <th class="collapsing center aligned" id="first-ignored">UNI</th>
```
[the remark this answers is not in the document]
Comment (typhainepl) [reply pronto/static/js/signatures/go.js:44]: that doesn't work, it still picks the attributes of the first child...

[the remark this answers is not in the document]
Comment (matthiasblum) [reply pronto/static/js/signatures/go.js:44]: Hmm... need to use the infamous `!important`:

Replace

https://github.com/ProteinsWebTeam/pronto/blob/4b9beb5c374b15af12b6675df82b48a1ed3412fe/pronto/static/js/signatures/go.js#L44

by

```js
                        <th class="collapsing center aligned normal">UNI</th>
```

Add the following in `pronto/static/css/go.css`:

```css
.ui.table > thead > tr > th.normal {
    border-left: 1px solid rgba(34, 36, 38, 0.1) !important;
    background: #F9FAFB !important;
    color: rgba(0, 0, 0, 0.87) !important;
    font-weight: bold !important;
}
```

Refresh

image

[the remark this answers is not in the document]
Comment (typhainepl) [reply pronto/static/js/signatures/go.js:44]: That worked, thanks!

File: pronto/api/signatures/go.py [hunk 3]
```diff
@@ -95,3 +102,72 @@ def get_go_terms(accessions):
...
+        )
+    for row in pg_cur:
+        go_id = row[0]
+        go_details[go_id] = {
+            "id":go_id,
+            "name": row[1],
+            "aspect": row[2],
+            }
+
+    return go_details
```
Comment (matthiasblum) [inline pronto/api/signatures/go.py:173]: PEP-8 & bind variables:

```suggestion
def get_go_details(term_ids: list[str], pg_cur) -> list[dict]:
    details = []

    binds = ["%s" for _ in term_ids]
    pg_cur.execute(
        f"""
        SELECT t.id, t.name, t.category
        FROM term t
        WHERE t.id IN ({','.join(binds)})
        """,
        term_ids
    )

    columns = ("id", "name", "aspect")
    for row in pg_cur.fetchall():
        details.append(dict(zip(columns, row)))

    return details

```

File: pronto/api/signatures/go.py [hunk 4]
```diff
@@ -95,3 +102,72 @@ def get_go_terms(accessions):
...
+            s[2].add(row[0])
+
+    cur.close()
+    con.close()
+
+    for terms_chunk in chunks(list(go_terms), 100):
+        go_details = get_go_details(terms_chunk, pg_cur)
+        for go, details in go_details.items():
+            terms[go].update(details)
+    return terms
```
Comment (matthiasblum) [inline pronto/api/signatures/go.py:153]: Using `range()` and modified version of `get_go_details()`:

```suggestion
    go_terms = list(go_terms)
    for i in range(0, len(go_terms), 100):
        subset = go_terms[i:i+100]

        for term in get_go_details(subset, pg_cur):
            terms[term["id"]].update(term)

    return terms
```

**Review by matthiasblum — CHANGES_REQUESTED** [review]
Looks better. There are still a few things to change, especially regarding bind variables. But we're almost there! :muscle:

**Comment by typhainepl** [comment]
@matthiasblum I applied the suggested changes

**Pull request merged**